feat: Add rule no-constant-binary-expression (#15296)
eslint/eslint: docs/src/rules/no-constant-binary-expression.md

# no-constant-binary-expression

> 

Reviewer: ```suggestion

Disallows expressions where the operation doesn't affect the value.

```

We usually add the short description (basically the same text as `meta.docs.description`) as the first paragraph in the docs.

```js
// Programmers coming from a language where objects are compared by value might expect this to work:
const isEmpty = x === [];

// However, this will always result in `isEmpty` being `false`.
```

## Rule Details

This rule identifies `==` and `===` comparisons which, based on the semantics of the JavaScript language, will always evaluate to `true` or `false`.

It also identifies `||`, `&&` and `??` logical expressions which will either always or never short circuit.

Examples of **incorrect** code for this rule:

```js
/*eslint no-constant-binary-expression: "error"*/

const value1 = +x == null;

const value2 = condition ? x : {} || DEFAULT;

const value3 = !foo == null;

const value4 = new Boolean(foo) === true;

const arrIsEmpty = someObj === {};

const arrIsEmpty = someArr === [];
```

Examples of **correct** code for this rule:

```js
/*eslint no-constant-binary-expression: "error"*/

const value1 = x == null;

const value2 = (condition ? x : {}) || DEFAULT;

const value3 = !(foo == null);

const value4 = Boolean(foo) === true;

> const arrIsEmpty = Object.keys(someObj).length === 0;

Reviewer: ```suggestion
const objIsEmpty = Object.keys(someObj).length === 0;
```

Same as the previous.

# no-constant-binary-expression

Disallows expressions where the operation doesn't affect the value.

> Comparisons which will always evaluate to true or false and logical expressions (`||`, `&&`, `??`) which either always short circuit or never short circuit are both likely indications of programmer error.

Reviewer 2: ```suggestion
Comparisons which will always evaluate to true or false and logical expressions (`||`, `&&`, `??`) which either always short-circuit or never short-circuit are both likely indications of programmer error.
```

These errors are especially common in complex expressions where operator precedence is easy to misjudge. For example:

```js
// One might think this would evaluate as `x + (b ?? c)`:
const x = a + b ?? c;

// But it actually evaluates as `(a + b) ?? c`. Since `a + b` can never be null,
// the `?? c` has no effect.
```

> Additionally, we detect comparisons to newly constructed objects/arrays/functions/etc. In JavaScript, where objects are compared by reference, a newly constructed object can _never_ `===` any other value. This can be surprising for programmers coming from languages where objects are compared by value.

Reviewer 2: ```suggestion
Additionally, this rule detects comparisons to newly constructed objects/arrays/functions/etc. In JavaScript, where objects are compared by reference, a newly constructed object can _never_ `===` any other value. This can be surprising for programmers coming from languages where objects are compared by value.
```

```js
// Programmers coming from a language where objects are compared by value might expect this to work:
const isEmpty = x === [];

// However, this will always result in `isEmpty` being `false`.
```

## Rule Details

This rule identifies `==` and `===` comparisons which, based on the semantics of the JavaScript language, will always evaluate to `true` or `false`.

> It also identifies `||`, `&&` and `??` logical expressions which will either always or never short circuit.

Reviewer 2: ```suggestion
It also identifies `||`, `&&` and `??` logical expressions which will either always or never short-circuit.
```
